### PR TITLE
crm_attribute: allow showing all node attributes

### PR DESCRIPTION
**Pull request by jnpkrn**
This is WIP, not to be merged yet.

### CODE REVIEW

File: lib/cib/cib_attrs.c [hunk 1]
```diff
@@ -86,10 +93,12 @@ find_nvpair_attr_delegate(cib_t * the_cib, const char *attr, const char *section
     }
 
     xpath_string = calloc(1, xpath_max);
-    offset += snprintf(xpath_string + offset, xpath_max - offset, "%.128s", get_object_path(section));
```
Comment (beekhof) [inline lib/cib/cib_attrs.c]: I like the old way TBH

Comment (jnpkrn) [inline lib/cib/cib_attrs.c]: Are we discussing aesthetics or a sound code?

```
$ CIB_file=cib-0733c987.xml valgrind crm_attribute --query -i "$(python -c 'print "X" * 960')"
[...]
==2701== Invalid write of size 1
==2701==    at 0x30BEA771B2: vsnprintf (vsnprintf.c:117)
==2701==    by 0x30BEA540D1: snprintf (snprintf.c:33)
==2701==    by 0x3016C10446: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
==2701==  Address 0x5398e43 is 3 bytes after a block of size 1,024 alloc'd
==2701==    at 0x4A08987: calloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==2701==    by 0x3016C102D5: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
==2701== 
==2701== Invalid write of size 1
==2701==    at 0x30BEA7B92B: _IO_default_xsputn (genops.c:475)
==2701==    by 0x30BEA496CC: vfprintf (vfprintf.c:1323)
==2701==    by 0x30BEA771C8: vsnprintf (vsnprintf.c:119)
==2701==    by 0x30BEA540D1: snprintf (snprintf.c:33)
==2701==    by 0x3016C10446: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
==2701==  Address 0x5398e43 is 3 bytes after a block of size 1,024 alloc'd
==2701==    at 0x4A08987: calloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==2701==    by 0x3016C102D5: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
==2701== 
==2701== Invalid write of size 1
==2701==    at 0x30BEA771DC: vsnprintf (vsnprintf.c:122)
==2701==    by 0x30BEA540D1: snprintf (snprintf.c:33)
==2701==    by 0x3016C10446: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
==2701==  Address 0x5398e44 is 4 bytes after a block of size 1,024 alloc'd
==2701==    at 0x4A08987: calloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==2701==    by 0x3016C102D5: find_nvpair_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x3016C115D6: read_attr_delegate (in /usr/lib64/libcib.so.4.0.0)
==2701==    by 0x4015CA: ??? (in /usr/sbin/crm_attribute)
==2701==    by 0x30BEA2078F: (below main) (libc-start.c:289)
[...]
```
Macro definition can be changed to resemble the original code more, though.

Comment (jnpkrn) [inline lib/cib/cib_attrs.c]: Plus with a macro in place, locally hardcoded (and only partially used) ".128" truncation can be dropped.

Comment (kgaillot) [inline lib/cib/cib_attrs.c]: It needs the memory fix, but I agree the old way was more readable. I see two alternative approaches: define the macro to just the snprintf, so it looks like:

offset += attr_snprintf(xpath_string, offset, xpath_max, "//%s", node_type);

Or define a struct with xpath_string, offset and xpath_max as members, so it looks like:

attr_snprintf(xpath_builder, "//%s", node_type);

The first appoach makes it more obvious that offset is being modified, and the second approach builds on yours to put the focus on the pattern rather than a list of variable names.

Comment (beekhof) [inline lib/cib/cib_attrs.c]: I would have thought that any attribute name over 128 would have been truncated and nothing would be written if `xpath_max - offset < 0`.

It seems `snprintf()` is unconditionally writing `\0` :-(

Comment (beekhof) [inline lib/cib/cib_attrs.c]: Better call it `crm_snprintf_max` and put it in include/crm_internal.h because we use that pattern all over the place.

Comment (jnpkrn) [inline lib/cib/cib_attrs.c]: Working on it, will send it as a separate PR.

Comment (jnpkrn) [inline lib/cib/cib_attrs.c]: Btw. `path_max - offset < 0` will never hold true when cast to unsigned type (`size_t`).

**Comment by jnpkrn** [comment]
Note that `snprintf` usage overhaul is a main subject of PR #810.

**Comment by jnpkrn** [comment]
Re https://github.com/ClusterLabs/pacemaker/pull/792#commitcomment-13191796:
I dunno what's better for internal API (here it would be `static` functions and `struct`s and `enums` used solely by them). My incilnation is towards having it in the header files, where it is usually found for the public API (here it would mean what the tools or daemons call directly). Having the parameters and return values of exported functions explained at actual function definitions feels like a very alternative approach.

**Comment by kgaillot** [comment]
There are arguments on both sides. See http://stackoverflow.com/questions/355619/where-to-put-the-doxygen-comment-blocks-for-an-internal-library-in-h-or-in-cpp

Doxygen will create public API documentation for use on website regardless, so the choice is more about what's easier for us to follow. I tend to find myself looking at .c's rather than .h's when tracing through issues.

**Comment by jnpkrn** [comment]
One more point, IIRC, parameter names in the function declaration not necessarily match the counterparts in its definition.
